### ISSUE TITLE
MacOS compatibility 

### ISSUE THREAD

**Issue by abu-sbeit** [Yellow-Dog-Man/Resonite-Issues#1412]
### Is your feature request related to a problem? Please describe.

I'm new to the VR building and programming, I'm MacOS user and currently Resonite doesn't work on MacOS.

<img width="504" alt="image" src="https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/115742964/6c00e3c4-3d06-48ed-bf33-590db42302c6">

### Describe the solution you'd like

Make it possible to work on MacOS as well

### Describe alternatives you've considered

Currently I'm using different software, but it doesn't provide all the features of Resonite

**Comment by Frooxius** [comment]
This is relatively big item, with lots of ongoing maintenance work to keep the builds functional over time, so we'd need to see how much interest there is in this to prioritize this and when.

The main things we need to do here on software side is:
- Make sure all our native dependencies are compiled for OS X
- Make sure that native texture uploads and other mechanisms work on OS X (which would likely use OpenGL)

There are two paths here:
1) If there's significant interest, we could prioritize this while still running on Unity. However I do not think we could support M1 and M2 with this, since the version of Unity we're not doesn't support that
2) We wait for the New Rendering Engine first, since this would avoid redoing some of the work needed for this and would open new options - https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1401

There is one part where community can help with this - if anyone's interested in contributing CI/CD for our native dependencies to make OS X dynamic libraries, that would help speed some of this up and make this a bit smaller task on our end.

**Comment by Geenz** [comment]
I’ve also done some testing on macOS using Rosetta- it’s not a fan of the version of Unity we’re on more generally. It’s something I want to do with the new rendering engine (and I actively use macOS for development on it anyways).

**Comment by TayIorRobinson** [comment]
can we at least get the badge, i mean the linux users do! :^)

<img width="510" height="254" alt="Image" src="https://github.com/user-attachments/assets/c30a641d-655d-40a8-a74b-40ce206dbc0e" />

**Comment by eth0fox** [comment]
Also, something I forgot to mention, while the `FrooxEngine.Platform` enum has a `OSX`, it appears that in StandaloneSystemInfo the logic is just `IsLinux ? Linux : Windows` (and since `IsLinux` is just checking if it's any *NIX like operating system, it will always mistake macOS for Linux instead of using something like [`System.OperatingSystem.IsMacOS()`](https://learn.microsoft.com/en-us/dotnet/api/system.operatingsystem.ismacos?view=net-10.0))

**Comment by Frooxius** [comment]
The enum entry is just for future purposes. 

There' no point in modifying checking this when we don't support running OS X in the first place. Our first priority will be making it work on high level.

**Comment by eth0fox** [comment]
Yeah, makes sense. It's effects are mostly cosmetic & It's not too hard to fix with mods anyway (or providing your own SystemInfo class)

**Comment by Frooxius** [comment]
We just fully open-sourced our renderer!

This makes it possible for community contributions towards this support if anyone is interested helping us out with this: https://github.com/Yellow-Dog-Man/Renderite.Unity.Renderer

**Comment by eth0fox** [comment]
There are quite a few things the need to be patched,

 - All the natives required by FrooxEngine need to be recompiled for macOS ARM64 as previously stated
 - The interprocess library needs to stop assuming because we're on UNIX we have `/dev/shm`: https://github.com/Yellow-Dog-Man/interprocess/pull/6
     -  Still not convinced this is a stellar idea as it seems that the contents of the file still seem to be persisted on disk, even though I don't think we're calling sync anywhere. (see previous comment from me)
 - Resonite still makes a bunch of assumptions that it's running in Win32 (either real, or under Wine) that I had to patch.
     - Renderite.Unity calls into user32.dll & CSCore's `WasapiOut`: https://github.com/eth0fox/ydms-Renderite/commit/1bbdaad7fe624fed825d939aec6e488bc5f298c8
     -  FrooxEngine itsself has a `Platform.OSX` enum value (which I've been using to ensure I don't get any Windows/Linux specific hacks), but StandaloneSystemInfo's autodetection logic treats any UNIX as Linux. Pass a custom SystemInfo such as https://github.com/eth0fox/resonite-osx/blob/main/OSXSystemInfo.cs
     - ~~FrooxEngine assumes the renderer lives at `Renderer/Renderite.Renderer.exe`, you can either put a shell script here to redirect the launch to the actual renderer, or patch the method.~~ (I assume this got fixed in 2026.5.11.1342)
 -  macOS uses Metal, not DX11, and Metal shaders Resonite doesn't have, so world loading will fail as soon as a StaticShader asset tries to load, (I see your `System.NotSupportedException: Unsupported engine platform: OSX`, but you can't stop a crazed fox that easily), so you have to patch the `Shader` class to immediately bail out loading.
   - If you just drag the Shaders repo into the Unity project, patching the engine to try and load the DX11 shaders, and then patching the Unity project to try and find the shaders in the local project instead, you can get _Some_ of the shaders to load 
 -  Texture loading is also broken for some reason. It appears the assumed size of the texture is bigger than Unity is actually allocating. Not sure. Haven't really looked into it and just patched [it](https://github.com/Yellow-Dog-Man/Renderite/blob/9c4c58d46a77a1e7a44b731db4067f8fab392421/Renderite.Unity/Assets/Texture/Texture2DAsset.cs#L147) to `i < Math.Min(texData.Length - offset, rawData.Length)` instead.  <img width="1106" height="349" alt="Image" src="https://github.com/user-attachments/assets/6163beef-1ade-4091-b201-16945cfa5cd2" />

But, if you do all that, and probably with a bit of luck, you can, get to see what resembles An Resonite

<img width="1912" height="1237" alt="Image" src="https://github.com/user-attachments/assets/bc841461-3afd-4a3c-82a7-eee5d0cefbd6" />


<details>
<summary>Other patches</summary>

```cs

// Disable randomisation of SHM file path for easier development
[HarmonyPatch(typeof (RenderiteMessagingHost), nameof(RenderiteMessagingHost.GenerateQueueName))]
class NoRandomSHMNamePatch {
    static bool Prefix (ref string __result) {
        __result = Engine.Current.SharedMemoryPrefix;
        return false;
    }
}

// Disable FrooxEngine trying to get the bootstrapper to launch the renderer instead of doing it directly
[HarmonyPatch(typeof(Engine), "IsRunningWithBootstrapper", MethodType.Getter)]
class IsRunningWithBootstrapperPatch {
    static bool Prefix(ref bool __result) {
        __result = false;
        return false;
    }
}


// Change the path FrooxEngine expects the renderer to live at
[HarmonyPatch(typeof (RenderSystem), "StartRenderer")]
class RendererPathPatch {
    static bool Prefix (ref RenderSystem __instance) {
        // __instance.DebugFramePacing = true;
        var field = typeof (RenderSystem).GetField("<RendererPath>k__BackingField", BindingFlags.Instance | BindingFlags.NonPublic);
        field.SetValue(__instance, "/Users/fox/dev/resonite-osx/Renderite.Renderer/Resonite.Renderer.app/Contents/MacOS/Renderite.Renderer");
        return true;
    }
}

// Disable shader loading entirely as Resonite doesn't have Metal shaders and will crash the world if you try.
[HarmonyPatch(typeof(Shader), "LoadTargetVariant")]
class ShaderDisablePatch {
    static bool Prefix(ref Task __result, ref Shader __instance) {
        
        var method = typeof (Shader).GetMethod("FailLoad", BindingFlags.Instance | BindingFlags.NonPublic);
        method.Invoke(__instance, new object?[] { "yeah no shaders here mate" });
        __result = Task.Delay(1);
        return false;
    }
}



[HarmonyPatch(typeof(Shader), "ShaderPlatform", MethodType.Getter)]
class ShaderPlatformPatch {
    static bool Prefix(ref ShaderTargetPlatform __result) {
        __result = ShaderTargetPlatform.WindowsDX11;
        return false;
    }
}
```


patch for https://github.com/Yellow-Dog-Man/Renderite/blob/master/Renderite.Unity/Assets/Shader/ShaderAsset.cs#L70

```cs
                                varvar shaderNameIdx = shaderRequest.asset.name.IndexOf("_");
                                var shaderName = shaderRequest.asset.name.Substring(0, shaderNameIdx);
                                var foundShader = Shader.Find(shaderName);
                                UnityEngine.Debug.Log("Found shader with name '" + shaderName + "' " + foundShader);
                                UnityShader = foundShader; //shaderRequest.asset as UnityEngine.Shader;
```

</details>
